astropy/units/quantity.py

```diff
@@ -685,7 +683,7 @@ def isscalar(self):
             (e.g. ``np.array(1)``), while this is True for quantities,
             since quantities cannot represent true numpy scalars.
         """
-        return not isiterable(self.value)
+        return not self.shape
```

Comment:
I suppose it's slower, but could you not just do `not self.shape and not self.dtype.fields` here to get around the "structured array is treated as a scalar" issue?